core/domain/acl_decorators.py

```diff
@@ -72,40 +71,286 @@ def check_activity_accessible(
             activity_rights.viewable_if_private)
 
 
-def check_exploration_editable(user_id, user_actions, exploration_id):
-    """Returns a boolean to signify whether given exploration is editable
-    by the user or not.
+def check_can_edit_activity(
+        user_id, user_actions, activity_type, activity_id):
+    """Checks whether the user can edit given activity.
 
     Args:
         user_id: str. Id of the given user.
-        user_actions: list(str). List of actions given user can perform.
-        exploration_id: str. Exploration id.
+        user_actions: list(str). List of actions the user can perform.
+        activity_id: str. Id of the given activity.
 
     Returns:
-        bool. Whether the given exploration can be accessed.
+        bool. Whether the given user can edit this activity.
     """
+    if user_id is None:
+        return False
+
+    activity_rights = (
+        rights_manager.get_exploration_rights(
+            activity_id, strict=False)
+        if activity_type == feconf.ACTIVITY_TYPE_EXPLORATION
+        else rights_manager.get_collection_rights(activity_id, strict=False))
+
+    if activity_rights is None:
```

Comment:
I'm not sure this is the place to do this. Shouldn't 404 instead be shown by the controller if the page doesn't exist? If in MapReduce processing I wanted to check who can edit an activity, a 404 should not be raised.

Basically, nothing but controllers and their decorators should be raising user-facing exceptions.